ext/background/background.js

```diff
@@ -171,16 +172,42 @@ async function attemptRecordInteraction(tab, session, force) {
     return Promise.resolve();
   }
   setChromeActivity(a, tab);
+
+  const merged = await mergeAttachment(session, a);
   return recordInteraction(
     {
       interaction_type: (_) => 'VIEWED',
       idempotency_key: `${moment().startOf('day')}::${tab.title}`,
-      attachment: a,
+      attachment: merged,
     },
     authorize(session)
   );
 }
 
+// Queries Range for existing attachments and merges the attachments based on a
+// provider's configured behavior. The Range backend will overwrite all fields
+// sent to it, but leave the rest alone.
+async function mergeAttachment(session, attachment) {
+  const dedupe = getProviderDedupe(attachment.provider);
+  if (dedupe == MERGE_BEHAVIOR.REPLACE_EXISTING) return attachment;
+
+  // This is the default, KEEP_EXISTING case
+  const activity = await listActivity(attachment.provider, authorize(session));
+  for (const a of activity.attachments) {
+    if (a.source_id != attachment.source_id) continue;
+
+    // Delete the properties that we want to preserve in the existing attachment
+    for (const f in a) {
+      // Send the core properties to ensure that the attachment is associated
+      // and updated properly in Range
+      if (ATTACHMENT_CORE.includes(f)) continue;
+      delete attachment[f];
+    }
```

Comment:
Once we find a matching attachment, we can exit out of the loop through `activity.attachments`, right?